In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import statsmodels.api as sm
from sklearn.linear_model import LinearRegression

%matplotlib widget

## 1

The p-values correspond to the statistical significance of the predictors (i.e. TV, radio, newspaper) of the simple linear regression. Based on the p-values, it can be concluded that TV and radio are significant, while newspaper is not statistically significant.

## 2 knn classifer vs. regressor

They are quite similar, except for the final step. In knn classifer, final prediction is made based on the majority class of the k-nearest neighbors, whereas in knn regressor, the response is the average of the responses of the k nearest neighbors.

## 3

### 3.a
The linear regression equation is (b0 + (b1 * gpa) + (b2 * iq) + (b4 * gpa * iq)) for males and (b0 + (b1 * gpa) + (b2 * iq) + (b4 * gpa * iq) + (b3 * 1) + (b5 * 1 * gpa). The delta b/w the two (females - males) is b3 + b5 * gpa = 35 - 10 * gpa. So, provide gpa is high enough, males CAN make more on avg. than females. So the correct option is (iii)

In [2]:
### 3.b
50 + 20 * 4 + 0.07 * 110 + 35 * 1 + 0.01 * 4 * 110 + -10 * 4 * 1

137.1

### 3.c
It is possible, but to be certain we need to look at t-statistic and p-values. 

## 4 training vs. test RSS

### 4.a
We expect training rss of cubic regression to be lower than training rss of linear regression (because cubic regression fits to the noise in data)

### 4.b
we expect test rss of cubic regresssion to be HIGHER than test rss of linear regression

### 4.c
We expect the training rss of cubic regression to be LOWER than linear regression irrespective of the nature of the true model.

### 4.d
We can't predict whether cubic or linear regession will have smaller test rss without knowing the true model. 

### 5
ai = 1 (?)

## 8 Simple linear regression on the Auto data set

In [3]:
auto = pd.read_csv('../data/Auto.csv')
auto.head()

mpg  cylinders  displacement horsepower  weight  acceleration  year  \
0  18.0          8         307.0        130    3504          12.0    70   
1  15.0          8         350.0        165    3693          11.5    70   
2  18.0          8         318.0        150    3436          11.0    70   
3  16.0          8         304.0        150    3433          12.0    70   
4  17.0          8         302.0        140    3449          10.5    70   

   origin                       name  
0       1  chevrolet chevelle malibu  
1       1          buick skylark 320  
2       1         plymouth satellite  
3       1              amc rebel sst  
4       1                ford torino

In [4]:
auto.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 397 entries, 0 to 396
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           397 non-null    float64
 1   cylinders     397 non-null    int64  
 2   displacement  397 non-null    float64
 3   horsepower    397 non-null    object 
 4   weight        397 non-null    int64  
 5   acceleration  397 non-null    float64
 6   year          397 non-null    int64  
 7   origin        397 non-null    int64  
 8   name          397 non-null    object 
dtypes: float64(3), int64(4), object(2)
memory usage: 28.0+ KB


In [5]:
# horsepower shoule be numeric but is treated as object. Are there any null values?
auto.horsepower.isna().any()

False

In [6]:
try:
    auto.horsepower.astype('int')
except ValueError as e:
    print(e)

invalid literal for int() with base 10: '?'


In [7]:
auto = pd.read_csv('../data/Auto.csv', na_values='?').dropna()
auto.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 392 entries, 0 to 396
Data columns (total 9 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   mpg           392 non-null    float64
 1   cylinders     392 non-null    int64  
 2   displacement  392 non-null    float64
 3   horsepower    392 non-null    float64
 4   weight        392 non-null    int64  
 5   acceleration  392 non-null    float64
 6   year          392 non-null    int64  
 7   origin        392 non-null    int64  
 8   name          392 non-null    object 
dtypes: float64(4), int64(4), object(1)
memory usage: 30.6+ KB


### 8.a

In [8]:
X8 = sm.add_constant(auto.horsepower)
y8 = auto.mpg
lm8 = sm.OLS(y8, X8).fit()
lm8.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.606
Model:                            OLS   Adj. R-squared:                  0.605
Method:                 Least Squares   F-statistic:                     599.7
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           7.03e-81
Time:                        17:14:12   Log-Likelihood:                -1178.7
No. Observations:                 392   AIC:                             2361.
Df Residuals:                     390   BIC:                             2369.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         39.9359      0.717     55.660      0.000      38.525      41.347
horsepower    -0.1578      0.006    -24.489      0.000      -0.171      -0.145
==============================================================================
Omnibus:                       16.432   Durbin-Watson:                   0.920
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               17.305
Skew:                           0.492   Prob(JB):                     0.000175
Kurtosis:                       3.299   Cond. No.                         322.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

i. yes, there is a relationship b/w horsepower and mpg (based on t-statistic and p-values).
ii. mpg = -0.1578 * horsepower => for 10 increase in horsepower, mpg drops by 1.578 gallons.
iii. negative

In [9]:
# iv. mpg for a horsepower of 98
X_test = np.array([1, 98])
print(lm8.predict(X_test))

pred = lm8.get_prediction(X_test)
pred.summary_frame()

[24.46707715]


mean   mean_se  mean_ci_lower  mean_ci_upper  obs_ci_lower  \
0  24.467077  0.251262      23.973079      24.961075     14.809396   

   obs_ci_upper  
0     34.124758

### 8.b mpg vs. horsepower

In [10]:
import seaborn as sns
plt.figure()
sns.regplot(x='horsepower', y='mpg', data=auto, ci=None,
           scatter_kws={'s':8},
           line_kws={'color':'r'})
plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

### 8.c diagnostic plots

In [11]:
f, ax = plt.subplots(1, 1)
# residual plot
sns.regplot(x=lm8.predict(), y=lm8.resid, ci=None, order=2,
           line_kws={'color':'r'})
ax.set_xlabel('fitted')
ax.set_ylabel('residuals')
ax.set_title('residual plot for the regression of mpg ~ horsepower')

plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

## 9 Multiple linear regression on the Auto data set

### 9.a scatterplot matrix

In [12]:
sns.pairplot(auto);

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

### 9.b

In [13]:
fig, ax = plt.subplots()
sns.heatmap(auto.corr())

plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

### 9.c

In [14]:
X9 = sm.add_constant(auto.drop(columns=['name', 'mpg']))
y9 = auto.mpg
lm9 = sm.OLS(y9, X9).fit()
lm9.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.821
Model:                            OLS   Adj. R-squared:                  0.818
Method:                 Least Squares   F-statistic:                     252.4
Date:                Sat, 12 Dec 2020   Prob (F-statistic):          2.04e-139
Time:                        17:14:18   Log-Likelihood:                -1023.5
No. Observations:                 392   AIC:                             2063.
Df Residuals:                     384   BIC:                             2095.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const          -17.2184      4.644     -3.707      0.000     -26.350      -8.087
cylinders       -0.4934      0.323     -1.526      0.128      -1.129       0.142
displacement     0.0199      0.008      2.647      0.008       0.005       0.035
horsepower      -0.0170      0.014     -1.230      0.220      -0.044       0.010
weight          -0.0065      0.001     -9.929      0.000      -0.008      -0.005
acceleration     0.0806      0.099      0.815      0.415      -0.114       0.275
year             0.7508      0.051     14.729      0.000       0.651       0.851
origin           1.4261      0.278      5.127      0.000       0.879       1.973
==============================================================================
Omnibus:                       31.906   Durbin-Watson:                   1.309
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               53.100
Skew:                           0.529   Prob(JB):                     2.95e-12
Kurtosis:                       4.460   Cond. No.                     8.59e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 8.59e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

### 9.c.i
there is definitely a relationship b/w weight, year & origin with mpg. The relationship b/w cylinders, displacement, horsepower & acceleration with mpg isn't so clear. There seems to be multicollinearity going on. Let's take a look.

In [15]:
from statsmodels.stats.outliers_influence import variance_inflation_factor
factors = [variance_inflation_factor(X9.values, i) for i in range(len(X9.columns))]
vif = pd.DataFrame({'vif factors': factors,
                    'predictors': X9.columns})
vif

vif factors    predictors
0   763.557531         const
1    10.737535     cylinders
2    21.836792  displacement
3     9.943693    horsepower
4    10.831260        weight
5     2.625806  acceleration
6     1.244952          year
7     1.772386        origin

### 9.c.2
weight, origin, year have satistically significant relationship with the response (mpg)

### 9.c.iii
every year, mpg increases by 0.7508, on average. 

### 9.d diagnostic plots

In [16]:
from scipy.stats import zscore

fitted = lm9.predict()
residuals = lm9.resid
norm_residuals = zscore(residuals)

fig, ax = plt.subplots(2, 2, figsize=(14, 10))
# residual plot--residuals vs. fitted
sns.residplot(x=fitted, y=residuals, ax=ax[0,0], lowess=True)
ax[0,0].set_xlabel('fitted')
ax[0,0].set_ylabel('residuals')
ax[0,0].set_title('residual plot')

# scale-location plot--studentized residuals vs. fitted
stu_residuals = np.sqrt(np.abs(norm_residuals))
ax[1,0].scatter(fitted, stu_residuals)
sns.regplot(x=fitted, y=stu_residuals, ax=ax[1,0], lowess=True, scatter=False, ci=None)
ax[1,0].set_xlabel('fitted')
ax[1,0].set_ylabel('studentized residuals')
ax[1,0].set_title('scale-locatioin plot')

# leverage plot
leverage = lm9.get_influence().hat_matrix_diag
ax[1,1].scatter(leverage, stu_residuals)
sns.regplot(x=leverage, y=stu_residuals, ax=ax[1,1], lowess=True, scatter=False, ci=None)

plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

### 9.e

In [17]:
auto['meta'] = auto.cylinders * auto.horsepower * auto.weight * auto.displacement
x9e = sm.add_constant(auto.drop(columns=['name', 'mpg']))
lm9e = sm.OLS(auto.mpg, x9e).fit()
lm9e.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                    mpg   R-squared:                       0.865
Model:                            OLS   Adj. R-squared:                  0.862
Method:                 Least Squares   F-statistic:                     306.9
Date:                Sat, 12 Dec 2020   Prob (F-statistic):          2.13e-161
Time:                        17:14:18   Log-Likelihood:                -968.65
No. Observations:                 392   AIC:                             1955.
Df Residuals:                     383   BIC:                             1991.
Df Model:                           8                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const           -6.1919      4.163     -1.487      0.138     -14.377       1.994
cylinders        0.1074      0.287      0.375      0.708      -0.456       0.671
displacement    -0.0240      0.008     -3.144      0.002      -0.039      -0.009
horsepower      -0.1049      0.014     -7.296      0.000      -0.133      -0.077
weight          -0.0059      0.001    -10.388      0.000      -0.007      -0.005
acceleration    -0.0729      0.087     -0.836      0.403      -0.244       0.098
year             0.7623      0.044     17.172      0.000       0.675       0.850
origin           0.6231      0.253      2.466      0.014       0.126       1.120
meta          7.271e-09   6.54e-10     11.119      0.000    5.99e-09    8.56e-09
==============================================================================
Omnibus:                       42.668   Durbin-Watson:                   1.499
Prob(Omnibus):                  0.000   Jarque-Bera (JB):               88.043
Skew:                           0.602   Prob(JB):                     7.62e-20
Kurtosis:                       4.985   Cond. No.                     2.96e+10
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 2.96e+10. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

## 10 Carseats data set

In [18]:
carseats = pd.read_csv('../data/Carseats.csv', index_col=0)
carseats.head()

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
1   9.50        138      73           11         276    120       Bad   42   
2  11.22        111      48           16         260     83      Good   65   
3  10.06        113      35           10         269     80    Medium   59   
4   7.40        117     100            4         466     97    Medium   55   
5   4.15        141      64            3         340    128       Bad   38   

   Education Urban   US  
1         17   Yes  Yes  
2         10   Yes  Yes  
3         12   Yes  Yes  
4         14   Yes  Yes  
5         13   Yes   No

In [19]:
carseats.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 400 entries, 1 to 400
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Sales        400 non-null    float64
 1   CompPrice    400 non-null    int64  
 2   Income       400 non-null    int64  
 3   Advertising  400 non-null    int64  
 4   Population   400 non-null    int64  
 5   Price        400 non-null    int64  
 6   ShelveLoc    400 non-null    object 
 7   Age          400 non-null    int64  
 8   Education    400 non-null    int64  
 9   Urban        400 non-null    object 
 10  US           400 non-null    object 
dtypes: float64(1), int64(7), object(3)
memory usage: 37.5+ KB


In [20]:
# are there any null values
carseats.isna().any(axis=0)

Sales          False
CompPrice      False
Income         False
Advertising    False
Population     False
Price          False
ShelveLoc      False
Age            False
Education      False
Urban          False
US             False
dtype: bool

### 10.a

In [21]:
x10a = pd.get_dummies(carseats[['Price', 'Urban', 'US']], drop_first=True)
y = carseats.Sales
x10a = sm.add_constant(x10a)
lm10a = sm.OLS(y, x10a).fit()
lm10a.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.234
Method:                 Least Squares   F-statistic:                     41.52
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           2.39e-23
Time:                        17:14:18   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1863.
Df Residuals:                     396   BIC:                             1879.
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.0435      0.651     20.036      0.000      11.764      14.323
Price         -0.0545      0.005    -10.389      0.000      -0.065      -0.044
Urban_Yes     -0.0219      0.272     -0.081      0.936      -0.556       0.512
US_Yes         1.2006      0.259      4.635      0.000       0.691       1.710
==============================================================================
Omnibus:                        0.676   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.713   Jarque-Bera (JB):                0.758
Skew:                           0.093   Prob(JB):                        0.684
Kurtosis:                       2.897   Cond. No.                         628.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 10.b
NOTE: Sales are in thousands, where as no units are mentioned for Pirce and CompPrice.
I assume Price to be in dollars. 
- For every 1 unit (dollar?) increase in price, sales drop by 54.5 units.
- Urban areas have 21.9 units of LESS sales than rural areas.
- US has 1200.6 more unit sales than non-US.

### 10.c
Sales = 13.0435 - (0.0545 * Price) - (0.0219 * Urban_Yes) + (1.2006 * US_Yes) # sales are in 1000 of units

### 10.d
we can safely reject the null hypothesis for 'Price' and 'US_Yes'.

### 10.e

In [22]:
x10e = pd.get_dummies(carseats[['Price', 'US']], drop_first=True)
y = carseats.Sales
x10e = sm.add_constant(x10e)
lm10e = sm.OLS(y, x10e).fit()
lm10e.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.239
Model:                            OLS   Adj. R-squared:                  0.235
Method:                 Least Squares   F-statistic:                     62.43
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           2.66e-24
Time:                        17:14:18   Log-Likelihood:                -927.66
No. Observations:                 400   AIC:                             1861.
Df Residuals:                     397   BIC:                             1873.
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         13.0308      0.631     20.652      0.000      11.790      14.271
Price         -0.0545      0.005    -10.416      0.000      -0.065      -0.044
US_Yes         1.1996      0.258      4.641      0.000       0.692       1.708
==============================================================================
Omnibus:                        0.666   Durbin-Watson:                   1.912
Prob(Omnibus):                  0.717   Jarque-Bera (JB):                0.749
Skew:                           0.092   Prob(JB):                        0.688
Kurtosis:                       2.895   Cond. No.                         607.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### 10.f
The model in 10.e (after dropping 'Urban' predictor) fits as well as the model in 10.a

### 10.g

In [23]:
lm10e.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

### 10.h
There seems to be atleast 1 high leverage point.

In [24]:
leverage = lm10e.get_influence().hat_matrix_diag
plt.figure()
# plt.scatter(leverage, lm10e.resid, alpha=0.6, s=8)
sns.regplot(x=leverage, y=lm10e.resid, ci=None, scatter=True, lowess=True,
           line_kws={'color':'r'})
plt.xlabel('leverage')
plt.ylabel('residuals')
plt.title('residuals vs. leverage')
plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

## 11

In [25]:
np.random.seed(1)
x = np.random.randn(100)
y = 2 * x + np.random.randn(100)
plt.figure()
plt.scatter(x, y);

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

### 11.a

In [26]:
lm11a = sm.OLS(y, x).fit()
lm11a.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

### 11.b

In [27]:
lm11b = sm.OLS(x, y).fit()
lm11b.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

### 11.c
In both cases, t-statistic and p-values are same. However, SE in the second case is lower than the first. That PROBABLY is because the variance of the error terms in the second case is 1/2 of the variance in the first case. 

### 11.f

In [28]:
lm1 = sm.OLS(y, sm.add_constant(x)).fit()
lm2 = sm.OLS(x, sm.add_constant(y)).fit()
lm1.tvalues[1], lm2.tvalues[1]

(19.78258531562925, 19.782585315629255)

## 12

In [29]:
np.random.seed(42)
x = np.random.randn(100)
noise = np.random.randn(100)

### 12.c

In [30]:
y = -1 * x 
lm12cyx = sm.OLS(y, x).fit()
lm12cxy = sm.OLS(x, y).fit()
lm12cyx.params, lm12cxy.params

(array([-1.]), array([-1.]))

### 12.b

In [31]:
y = -1 * x + noise
lm12cyx = sm.OLS(y, x).fit()
lm12cxy = sm.OLS(x, y).fit()
lm12cyx.params, lm12cxy.params

(array([-1.14418953]), array([-0.481294]))

## 13

### 13 a, b, c

In [32]:
np.random.seed(1)
x = np.random.normal(loc=0.0, scale=1.0, size=100)
eps = np.random.normal(loc=0.0, scale=0.25, size=100)
y = -1 + 0.5 * x + eps

Length of y is 100. beta_0 = -1 and beta_1 = 0.5

### 13.d

In [33]:
plt.figure()
plt.scatter(x, y, s=12);

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

### 13.e

In [34]:
lm13e = sm.OLS(y, sm.add_constant(x)).fit()
lm13e.summary().tables[1]

<class 'statsmodels.iolib.table.SimpleTable'>

### 13.f

In [35]:
plt.figure()
plt.scatter(x, y, s=12)
plt.plot(x, (-1 + 0.5 *x), 'k-', label='Pop. regr. line')
plt.plot(x, (lm13e.params[0] + lm13e.params[1] * x), 'r--', label='Least squares regr. line')
plt.legend()
plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

### 13.g polynomial regression

In [36]:
X = pd.DataFrame({'x1': x, 'x2': x*x})
lm13g = sm.OLS(y, sm.add_constant(X)).fit()
print(lm13g.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.796
Method:                 Least Squares   F-statistic:                     193.8
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           1.32e-34
Time:                        17:14:19   Log-Likelihood:                 4.2077
No. Observations:                 100   AIC:                            -2.415
Df Residuals:                      97   BIC:                             5.400
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9663      0.029    -33.486      0.0

In [37]:
print(lm13e.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.798
Method:                 Least Squares   F-statistic:                     391.4
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           5.39e-36
Time:                        17:14:19   Log-Likelihood:                 4.1908
No. Observations:                 100   AIC:                            -4.382
Df Residuals:                      98   BIC:                            0.8288
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9632      0.023    -40.999      0.0

Looking at both the R2 values and t-statistic and p-values for the x2 term, polynomial regression doesn't improve over linear regression.

### 13.h

In [38]:
yLessNoise = -1 + 0.5 * x + (eps * 0.1)
plt.figure()
plt.scatter(x, yLessNoise, s=12)
plt.plot(x, (-1 + 0.5 * x), 'k-', label='True model')
lm13hLin = sm.OLS(yLessNoise, sm.add_constant(x)).fit()
print('Less noise linear fit:', lm13hLin.summary())
plt.plot(x, (lm13hLin.params[0] + lm13hLin.params[1] * x), 'r--', label='Less noise: linear fit')

plt.legend()
plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

Less noise linear fit:                             OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.997
Model:                            OLS   Adj. R-squared:                  0.997
Method:                 Least Squares   F-statistic:                 3.599e+04
Date:                Sat, 12 Dec 2020   Prob (F-statistic):          1.46e-127
Time:                        17:14:19   Log-Likelihood:                 234.45
No. Observations:                 100   AIC:                            -464.9
Df Residuals:                      98   BIC:                            -459.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9963      0.

### 13.i

In [39]:
yMoreNoise = -1 + 0.5 * x + (eps * 5) # 5 times more noise
plt.figure()
plt.scatter(x, yMoreNoise, s=12)
plt.plot(x, (-1 + 0.5 * x), 'k-', label='True model')
lm13iLin = sm.OLS(y, sm.add_constant(x)).fit()
print('More nose linear filt:\n', lm13iLin.summary())
plt.plot(x, (lm13iLin.params[0] + lm13iLin.params[1] * x), 'r--', label='More noise: linear fit')

lm13iPoly = sm.OLS(y, sm.add_constant(X)).fit()
print('\nMore noise poly fit:\n', lm13iPoly.summary())
plt.plot(x, (lm13iPoly.params[0] + lm13iPoly.params[1] * x + lm13iPoly.params[2] * x * x), 'b-.', label='More noise: poly fit')
plt.legend()
plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

More nose linear filt:
                             OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.800
Model:                            OLS   Adj. R-squared:                  0.798
Method:                 Least Squares   F-statistic:                     391.4
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           5.39e-36
Time:                        17:14:19   Log-Likelihood:                 4.1908
No. Observations:                 100   AIC:                            -4.382
Df Residuals:                      98   BIC:                            0.8288
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.9632      0

## 14 collinearity

### 14.a

In [40]:
np.random.seed(1)
x1 = np.random.uniform(size=100)
x2 = 0.5*x1 + np.random.standard_normal(100) * 0.1
y = 2  + 2*x1 + 0.3*x2 + np.random.standard_normal(100)
X = pd.DataFrame({"x1": x1, 'x2': x2})

### 14.b correlation b/w x1 & x2

In [41]:
print(X.corr())


         x1       x2
x1  1.00000  0.83906
x2  0.83906  1.00000


### 14.c

In [42]:
lm14c = sm.OLS(y, sm.add_constant(X)).fit()
print(lm14c.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.261
Model:                            OLS   Adj. R-squared:                  0.245
Method:                 Least Squares   F-statistic:                     17.09
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           4.40e-07
Time:                        17:14:19   Log-Likelihood:                -142.34
No. Observations:                 100   AIC:                             290.7
Df Residuals:                      97   BIC:                             298.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.1893      0.199     11.021      0.0

- Although the true model is linear and we are fitting a OLS linear model, R2 is not high. 
- One would expect R2 to be higher, if NOT for the collinearity b/w x1 & x2.
- beta estimates are 2.2, 0.7 & 2.5, whereas the true beta values are: 2, 2 & 0.3
- we can't reject either of the null hypothesis

### 14.d

In [43]:
lm14d = sm.OLS(y, sm.add_constant(x1)).fit()
print(lm14d.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.224
Model:                            OLS   Adj. R-squared:                  0.216
Method:                 Least Squares   F-statistic:                     28.26
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           6.68e-07
Time:                        17:14:19   Log-Likelihood:                -144.76
No. Observations:                 100   AIC:                             293.5
Df Residuals:                      98   BIC:                             298.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.2486      0.201     11.209      0.0

### 14.e

In [44]:
lm14e = sm.OLS(y, sm.add_constant(x2)).fit()
print(lm14e.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.251
Model:                            OLS   Adj. R-squared:                  0.244
Method:                 Least Squares   F-statistic:                     32.87
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           1.09e-07
Time:                        17:14:19   Log-Likelihood:                -142.97
No. Observations:                 100   AIC:                             289.9
Df Residuals:                      98   BIC:                             295.1
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.2655      0.187     12.145      0.0

**NOTE**: why doesn't the beta2 estimate (=3.5613) doesn't match true beta2 (=0.3)?

### 14.f
The results obtained above contradict each other wrt. null hypothesis. In 14.c, we couldn't reject the null hypothesis, whereas in 14.d & 14.e we can reject the null hypothesis with high confidence. When there is collinearity involved, the results obtained by the model with both x1 & x2 predictors cannot be trusted. 

We can probably imporve the fit adding a new predictor variable x3 = x1 * x2

In [45]:
X['x3'] = (X.x1 + X.x2) / 2 # mean
lm14f = sm.OLS(y, sm.add_constant(X)).fit()
print(lm14f.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.261
Model:                            OLS   Adj. R-squared:                  0.245
Method:                 Least Squares   F-statistic:                     17.09
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           4.40e-07
Time:                        17:14:19   Log-Likelihood:                -142.34
No. Observations:                 100   AIC:                             290.7
Df Residuals:                      97   BIC:                             298.5
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.1893      0.199     11.021      0.0

### 14.g

In [46]:
x1 = np.append(x1, 0.1)
x2 = np.append(x2, 0.8)
y = np.append(y, 6)
X = pd.DataFrame({'x1': x1, 'x2': x2})

plt.figure()
plt.scatter(x1, x2, s=12);

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

In [47]:
# linear fit using both the predictors
lm14g1 = sm.OLS(y, sm.add_constant(X)).fit()
print(lm14g1.summary())
print('\n')
print(lm14c.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.289
Model:                            OLS   Adj. R-squared:                  0.275
Method:                 Least Squares   F-statistic:                     19.92
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           5.51e-08
Time:                        17:14:20   Log-Likelihood:                -144.15
No. Observations:                 101   AIC:                             294.3
Df Residuals:                      98   BIC:                             302.1
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.1996      0.199     11.041      0.0

In [48]:
# linear fit using only x1
lm14g2 = sm.OLS(y, sm.add_constant(x1)).fit()
print(lm14g2.summary())
print('\n', lm14d.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.181
Model:                            OLS   Adj. R-squared:                  0.173
Method:                 Least Squares   F-statistic:                     21.89
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           9.15e-06
Time:                        17:14:20   Log-Likelihood:                -151.28
No. Observations:                 101   AIC:                             306.6
Df Residuals:                      99   BIC:                             311.8
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.3583      0.208     11.331      0.0

In [49]:
# linear fit using only x2
lm14g3 = sm.OLS(y, sm.add_constant(x2)).fit()
print(lm14g3.summary())
print('\n', lm14e.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.281
Method:                 Least Squares   F-statistic:                     39.99
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           7.41e-09
Time:                        17:14:20   Log-Likelihood:                -144.24
No. Observations:                 101   AIC:                             292.5
Df Residuals:                      99   BIC:                             297.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.2331      0.182     12.271      0.0

In [50]:
fig, ax = plt.subplots(1, 3, figsize=(16, 6))
# residuals vs. leverage plots
l1 = lm14g1.get_influence().hat_matrix_diag
ax[0].plot(l1, lm14g1.resid, 'bo', alpha=0.6)
ax[0].set_xlabel('leverage')
ax[0].set_ylabel('residuals')
ax[0].set_title('regression with both predictors x1 & x2')

l2 = lm14g2.get_influence().hat_matrix_diag
ax[1].plot(l2, lm14g2.resid, 'bo', alpha=0.6)
ax[1].set_xlabel('leverage')
ax[1].set_ylabel('residuals')
ax[1].set_title('regression with one predictors x1')

l3 = lm14g3.get_influence().hat_matrix_diag
ax[2].plot(l3, lm14g3.resid, 'bo', alpha=0.6)
ax[2].set_xlabel('leverage')
ax[2].set_ylabel('residuals')
ax[2].set_title('regression with one predictors x2')

plt.show();

Canvas(toolbar=Toolbar(toolitems=[('Home', 'Reset original view', 'home', 'home'), ('Back', 'Back to previous …

NOTES:
- the coefficients change quite a bit for the regression with both the predictors. The new observation is high leverage.
- the coefficients don't change at all for the regression with the predictor x1. It is not high leverage. It could probably be an outlier.
- the coefficients don't change much for the regression with the predictor x2. It could be high leverage, although it is NOT very clear. It is definitely not an outlier. 

## 15 Boston data set

In [51]:
boston = pd.read_csv('../data/Boston.csv', index_col=0, na_values='?').dropna()
boston.head()

crim    zn  indus  chas    nox     rm   age     dis  rad  tax  ptratio  \
1  0.00632  18.0   2.31     0  0.538  6.575  65.2  4.0900    1  296     15.3   
2  0.02731   0.0   7.07     0  0.469  6.421  78.9  4.9671    2  242     17.8   
3  0.02729   0.0   7.07     0  0.469  7.185  61.1  4.9671    2  242     17.8   
4  0.03237   0.0   2.18     0  0.458  6.998  45.8  6.0622    3  222     18.7   
5  0.06905   0.0   2.18     0  0.458  7.147  54.2  6.0622    3  222     18.7   

    black  lstat  medv  
1  396.90   4.98  24.0  
2  396.90   9.14  21.6  
3  392.83   4.03  34.7  
4  394.63   2.94  33.4  
5  396.90   5.33  36.2

In [52]:
boston.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 506 entries, 1 to 506
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   crim     506 non-null    float64
 1   zn       506 non-null    float64
 2   indus    506 non-null    float64
 3   chas     506 non-null    int64  
 4   nox      506 non-null    float64
 5   rm       506 non-null    float64
 6   age      506 non-null    float64
 7   dis      506 non-null    float64
 8   rad      506 non-null    int64  
 9   tax      506 non-null    int64  
 10  ptratio  506 non-null    float64
 11  black    506 non-null    float64
 12  lstat    506 non-null    float64
 13  medv     506 non-null    float64
dtypes: float64(11), int64(3)
memory usage: 59.3 KB


NOTES:
- chas is qualitative with two values. It is already encoded.
- rad can be treated as both qualitative and quantitative. Leave it as is. 
- all other predictors are quantitative. 

### 15.b multiple regression

In [53]:
X = boston.drop(columns='crim')
y = boston.crim
lm15_mul = sm.OLS(y, sm.add_constant(X)).fit()
print(lm15_mul.summary())

                            OLS Regression Results                            
Dep. Variable:                   crim   R-squared:                       0.454
Model:                            OLS   Adj. R-squared:                  0.440
Method:                 Least Squares   F-statistic:                     31.47
Date:                Sat, 12 Dec 2020   Prob (F-statistic):           1.57e-56
Time:                        17:14:20   Log-Likelihood:                -1653.3
No. Observations:                 506   AIC:                             3335.
Df Residuals:                     492   BIC:                             3394.
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         17.0332      7.235      2.354      0.0

In [54]:
# TODO